lib/timber/transports/http.ex

```diff
@@ -0,0 +1,189 @@
+  previous (first) request is still being processed, then the transport will enter
+  synchronous mode, waiting for a response before proceeding with the request.
+
+  ## Configuration
+
+  ### Custom HTTP client
+
+  By default, hackney is used via `Timber.Transports.HTTP.HackneyClient`. You can define your
+  own custom HTTP client by adhering to the `Timber.Transports.HTTP.Client` behaviour. Afterwards,
+  you must specify your client in the configuration:
+
+  ```
+  config :timber, :http_transport, http_client: MyHTTPClient
+  ```
+  """
+
+  @behaviour Timber.Transport
+
+  alias Timber.LogEntry
+
+  @typep t :: %__MODULE__{
+    api_key: String.t,
+    buffer_size: non_neg_integer,
+    buffer: [] | [IO.chardata],
+    flush_interval: non_neg_integer,
+    max_buffer_size: pos_integer,
+    ref: reference
+  }
+
+  @content_type "application/msgpack"
+  @default_max_buffer_size 5000 # 5000 log line should be well below 5mb
+  @default_flush_interval 1000
+  @default_http_client __MODULE__.HackneyClient
+  @url "https://api.timber.io/frames"
+
+  defstruct api_key: nil,
+            buffer_size: 0,
+            buffer: [],
+            flush_interval: @default_flush_interval,
+            max_buffer_size: @default_max_buffer_size,
+            ref: nil
+
+  @doc false
+  @spec init() :: {:ok, t}
+  def init() do
+    config = Keyword.put(config(), :api_key, Timber.Config.api_key())
+
+    case configure(config, %__MODULE__{}) do
+      {:ok, state} ->
+        flusher(state)
+        {:ok, state}
+      {:error, error} -> {:error, error}
+    end
+  end
+
+  @doc false
+  @spec configure(Keyword.t, t) :: {:ok, t}
+  def configure(options, %{api_key: current_api_key} = state) do
+    api_key = Keyword.get(options, :api_key, current_api_key)
+    max_buffer_size = Keyword.get(options, :max_buffer_size, @default_max_buffer_size)
+    new_state = %{ state | api_key: api_key, max_buffer_size: max_buffer_size }
+
+    if api_key == nil do
+      {:error, :no_api_key}
+    else
+      {:ok, new_state}
+    end
+  end
+
+  @doc false
+  @spec write(LogEntry.t, t) :: {:ok, t}
+  def write(log_entry, state) do
+    # Write to the buffer immediately because we want to batch lines and send
+    # them on an interval.
+    state = write_buffer(log_entry, state)
+
+    if state.buffer_size === state.max_buffer_size do
+      # The buffer is full, flush immediately.
+      {:ok, flush(state)}
+    else
+      {:ok, state}
+    end
+  end
+
+  # Writes a log entry into the buffer
+  @spec write_buffer(LogEntry.t, t) :: t
+  defp write_buffer(log_entry, %{buffer: buffer, buffer_size: buffer_size} = state) do
+    %{state | buffer: buffer ++ [log_entry], buffer_size: buffer_size + 1}
+  end
+
+  # Handle the flusher step, this recursively calls itself by re-calling `flusher/1`.
+  # This is how the flush interval is maintained.
+  @doc false
+  @spec handle_info(atom(), t) :: {:ok, t}
+  def handle_info(:flusher_step, state) do
```

Comment:
To make this step asynchronous. `flush/1` waits on the request to complete because of your previous comment about `Logger.flush/1`. In the odd case that a request takes longer than the flush interval, it will wait until the previous request is finished before sending the new one.